bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java

```diff
@@ -29,211 +34,315 @@
     /**
-     * Start service.
+     * Create broker connections based on the service configuration.
      */
+    @Override
+    public void updated(Dictionary<String, ?> properties) {
+        // load broker configurations from configuration file
+        if (properties == null || properties.isEmpty()) {
+            return;
+        }
+
+        Map<String, Map<String, Object>> brokerConfigs = extractBrokerConfigurations(properties);
+
+        for (Map<String, Object> brokerConfig : brokerConfigs.values()) {
+            try {
+                addBrokerConnection(brokerConfig).start();
+            } catch (ConfigurationException e) {
+                logger.warn("MqttBroker connection configuration faulty: {}", e.getMessage());
+            } catch (MqttException e) {
+                logger.warn("MqttBroker start failed: {}", e.getMessage(), e);
+            }
+        }
+    }
+
     public void activate() {
         logger.debug("Starting MQTT Service...");
     }
 
-    /**
-     * Stop service.
-     */
     public void deactivate() {
         logger.debug("Stopping MQTT Service...");
-        stopAllBrokerConnections();
-        logger.debug("MQTT Service stopped.");
+        for (final MqttBrokerConnection conn : brokerConnections.values()) {
+            conn.close();
+        }
+        brokerConnections.clear();
     }
 
-    private void startAllBrokerConnections() {
-        /*
-         * No need to synchronize the access to the broker connections.
-         * We don't add or remove entries only use the existing ones and the map is a concurrent one.
-         */
-        for (final MqttBrokerConnection conn : brokerConnections.values()) {
-            try {
-                conn.start();
-            } catch (final Exception e) {
-                conn.connectionLost(e);
-                logger.error("Error starting broker connection", e);
+    /**
+     * Add a listener to get notified of new/removed brokers.
+     *
+     * @param observer The observer
+     */
+    public void addBrokersListener(MqttBrokersObserver observer) {
+        brokersObservers.add(observer);
+    }
+
+    /**
+     * Remove a listener and don't get notified of new/removed brokers anymore.
+     *
+     * @param observer The observer
+     */
+    public void removeBrokersListener(MqttBrokersObserver observer) {
+        brokersObservers.remove(observer);
+    }
+
+    /**
+     * Lookup an broker connection by name.
+     *
+     * @param brokerName to look for.
+     * @return existing connection or null
+     */
+    public MqttBrokerConnection getBrokerConnection(String brokerName) {
+        synchronized (brokerConnections) {
+            return brokerConnections.get(brokerName.toLowerCase());
+        }
+    }
+
+    /**
+     *
+     * @param connection
+     */
+    public void addBrokerConnection(MqttBrokerConnection connection) {
+        synchronized (brokerConnections) {
+            brokerConnections.put(connection.getName(), connection);
```

Comment:
Is it guaranteed that externally created BrokerConnections also have an all lower case name? Otherwise you will not be able to get it back from the MqttService (see line 144).